Updating Native Modules Documentation (#2261)
facebook/react-native-website: website/versioned_docs/version-0.63/native-modules-android.md

We also recommend enabling [Gradle Daemon](https://docs.gradle.org/2.9/userguide/gradle_daemon.html) to speed up builds as you iterate on Java code.

### Create A Custom Native Module File

The first step is to create a main custom native module file. Create a new Java class named CalendarModule.java inside `android/app/src/main/java/com/your-app-name/ folder`.

<figure>
  <img src="/docs/assets/native-modules-android-add-class.png" width="700" alt="Image of adding a class called CalendarModule.java within the Android Studio." />
  <figcaption>Image of how to add the CalendarModuleClass</figcaption>
</figure>

Then add the following content:

```jsx
package com.your-app-name; // replace com.your-app-name with your app’s name
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import java.util.Map;
import java.util.HashMap;

public class CalendarModule extends ReactContextBaseJavaModule {
   CalendarModule(ReactApplicationContext context) {
       super(context);
   }
}
```

> As you can see, your `CalendarModule` class extends the `ReactContextBaseJavaModule`. For Android, most native modules consist of a class that extends `ReactContextBaseJavaModule` and implements the functionality required by Javascript.

Reviewer: The word "consist", how it's used here, is a bit confusing: "For Android, most native modules consist of a class". 

```suggestion
As you can see, your `CalendarModule` class extends the `ReactContextBaseJavaModule` class. For Android, Java native modules are written as classes that extend `ReactContextBaseJavaModule` and implement the functionality required by Javascript.
```

> > Note: It is worth noting that technically Java classes can extend the `BaseJavaModule` class or implement the `NativeModule` interface to be considered a Native Module by React Native.

Reviewer: ```suggestion
> Note: It is worth noting that technically Java classes only needs to extend the `BaseJavaModule` class or implement the `NativeModule` interface to be considered a Native Module by React Native.
```

Author: Applying this in a separate commit to fix a grammar issue! (should be: only need to)

> However we recommend that you use `ReactContextBaseJavaModule`, as shown above. `ReactContextBaseJavaModule` gives access to the `ReactApplicationContext` (RAC), which is useful for Native Modules that need to hook into activity lifecycle methods. Using `ReactContextBaseJavaModule` will also make it easier to make your native module type-safe in the future. For native module type-safety, which is coming in future releases, React Native looks at each native module's Javascript spec and generates an abstract base class that extends `ReactContextBaseJavaModule`.

### Module Name

> All native modules in Android need to implement the `getName()` method. This method returns the string name of the native module which represents the class in JavaScript. The native module can then be accessed in JS via its name. For example, in the below code snippet, `getName` returns `"CalendarModule"`.

Reviewer: ```suggestion
All Java native modules in Android need to implement the `getName()` method. This method returns a string, which represents the name of the native module. The native module can then be accessed in JavaScript using its name. For example, in the below code snippet, `getName()` returns `"CalendarModule"`.
```

}
```

The native module can then be accessed in JS like this:

```jsx
const { CalendarModule } = ReactNative.NativeModules;
```

### Export a Native Method to JavaScript

Next you will need to add a method to your native module that will create calendar events and can be invoked in Javascript. All native module methods meant to be invoked from JavaScript must be annotated with `@ReactMethod`.

Set up a method `createCalendarEvent` for `CalendarModule` that can be invoked in JS through `CalendarModule.createCalendarEvent()`. For now, the method will take in a name and location as strings. We will cover parameters options shortly.

```java
@ReactMethod
public void createCalendarEvent(String name, String location) {
}
```

Add a debug log in the method to confirm it has been invoked when you call it from your application. Below we import and use the [Log](https://developer.android.com/reference/android/util/Log) class from the Android util package like so:

```java
import android.util.Log;

@ReactMethod
public void createCalendarEvent(String name, String location) {
   Log.d("CalendarModule", "Create event called with name: " + name
   + " and location: " + location);
}
```

> Once we finish implementing the native module and hook it up in JS, you can follow [these steps](https://developer.android.com/studio/debug/am-logcat.html) to view the logs from your app.

Reviewer: ```suggestion
Once we finish implementing the native module and hook it up in JavaScript, you can follow [these steps](https://developer.android.com/studio/debug/am-logcat.html) to view the logs from your app.
```

### Synchronous Methods

You can pass `isBlockingSynchronousMethod = true` to a native method to mark it as a synchronous method.

```java
@ReactMethod(isBlockingSynchronousMethod = true)
```

At the moment, we do not recommend this, since calling methods synchronously can have strong performance penalties and introduce threading-related bugs to your native modules. Additionally, please note that if you choose to enable `isBlockingSynchronousMethod`, your app can no longer use the Google Chrome debugger. This is because synchronous methods require the JS VM to share memory with the app. For the Google Chrome debugger, React Native runs inside the JS VM in Google Chrome, and communicates asynchronously with the mobile devices via WebSockets.

### Register the Module (Android Specific)

> Once a native module is written, it needs to be registered with React Native. In order to do so, you need to add your native module to a `ReactPackage` and register the `ReactPackage` with React Native. During initialization, React Native will loop over all packages, and for each `ReactPackage`, register each native module within.

Reviewer: > register each native module within.

After reading this, I'm left wondering what it means to register a NativeModule with React Native. Maybe we should elaborate on this. What we're really doing is making each NativeModule available to JavaScript code.

**Edit:** Looks like we already talk about this down the line.

import java.util.Collections;
import java.util.List;

public class MyAppPackage implements ReactPackage {

   @Override
   public List<ViewManager> createViewManagers(ReactApplicationContext reactContext) {
       return Collections.emptyList();
   }

   @Override
   public List<NativeModule> createNativeModules(
           ReactApplicationContext reactContext) {
       List<NativeModule> modules = new ArrayList<>();

       modules.add(new CalendarModule(reactContext));

       return modules;
   }

}
```

This file imports the native module we created, `CalendarModule`. It then instantiates `CalendarModule` within the `createNativeModules` function and returns it as a list of `NativeModules` to register. If you add more native modules down the line, you can also instantiate them and add them to the list returned here.

> > Note: It is worth noting that this way of registering native modules eagerly initializes all native modules when the application starts, which adds to the startup time of an application. You can use [TurboReactPackage](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/TurboReactPackage.java) as an alternative. Instead of `createNativeModules`, which return a list of instantiated native module objects, TurboReactPackage implements a `getModule(String name, ReactApplicationContext rac)` method that creates the native module object, when required. TurboReactPackage is a bit more complicated to implement at the moment. In addition to implementing a `getModule` method, you have to implement a `getReactModuleInfoProvider()` method, which returns a list of all the native modules the package can instantiate along with a function that instantiates them, example [here](https://git.io/JfuT0). Again, using TurboReactPackage will allow your application to have a faster startup time, but it is a bit more difficult to write and maintain. So proceed with caution if you choose to use TurboReactPackages.

Reviewer: Might be a good idea to just use the full GitHub links, as opposed to redirects.

```suggestion
> Note: It is worth noting that this way of registering native modules eagerly initializes all native modules when the application starts, which adds to the startup time of an application. You can use [TurboReactPackage](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/TurboReactPackage.java) as an alternative. Instead of `createNativeModules`, which return a list of instantiated native module objects, TurboReactPackage implements a `getModule(String name, ReactApplicationContext rac)` method that creates the native module object, when required. TurboReactPackage is a bit more complicated to implement at the moment. In addition to implementing a `getModule` method, you have to implement a `getReactModuleInfoProvider()` method, which returns a list of all the native modules the package can instantiate along with a function that instantiates them, example [here](https://github.com/facebook/react-native/blob/8ac467c51b94c82d81930b4802b2978c85539925/ReactAndroid/src/main/java/com/facebook/react/CoreModulesPackage.java#L86-L165). Again, using TurboReactPackage will allow your application to have a faster startup time, but it is currently a bit cumbersome to write. So proceed with caution if you choose to use TurboReactPackages.
```

To register the `CalendarModule` package, you must add `MyAppPackage` to the list of packages returned in ReactNativeHost's `getPackages` method. Open up your `MainApplication.java` file, which can be found in the following path: `android/app/src/main/java/com/your-app-name/MainApplication.java`

Locate ReactNativeHost’s `getPackages` method and add your package to the packages list `getPackages` returns:

```java
@Override
  protected List<ReactPackage> getPackages() {
    @SuppressWarnings("UnnecessaryLocalVariable")
    List<ReactPackage> packages = new PackageList(this).getPackages();
    // below we add MyAppPackage to the list of packages returned
    packages.add(new MyAppPackage());
    return packages;
  }
```

You have now successfully registered your native module for Android!

### Test What You Have Built

> At this point, you have set up the basic scaffolding for your native module in Android. Test that out by accessing the native module and invoking it’s exported method in JavaScript.

Reviewer: ```suggestion
At this point, you have set up the basic scaffolding for your native module in Android. Test that out by accessing the native module and invoking its exported method in JavaScript.
```

Find a place in your application where you would like to add a call to the native module’s `createCalendarEvent` method. In the following example we created a component, NewModuleButton, and will invoke the native module inside the onPress function.

```jsx
import React from 'react';
import { NativeModules, Button } from 'react-native';

const NewModuleButton = () => {
  const onPress = () => {
    console.log('We will invoke the native module here!');
  };

  return (
    <Button
      title="Click to invoke your native module!"
      color="#841584"
      onPress={onPress}
    />
  );
};

export default NewModuleButton;
```

> In order to access your native module from Javascript you need to first import `NativeNodules` from React Native:

Reviewer: ```suggestion
In order to access your native module from Javascript you need to first import `NativeModules` from React Native:
```

Author: Nodules!! Gah good catch 🤪

```java
@Override
public Map<String, Object> getConstants() {
   final Map<String, Object> constants = new HashMap<>();
   constants.put("DEFAULT_EVENT_NAME", "New Event");
   return constants;
}
```

The constant can then be accessed by invoking `getConstants` on the native module in JS:

```jsx
const { DEFAULT_EVENT_NAME } = CalendarModule.getConstants();
console.log(DEFAULT_EVENT_NAME);
```

Technically it is possible to access constants exported in `getConstants` directly off the native module object. This will no longer be supported with TurboModules, so we encourage the community to switch to the above approach to avoid necessary migration down the line.

> Note that currently constants are exported only at initialization time, so if you change getConstants values at runtime it won't affect the JavaScript environment. This will change with Turbomodules. With Turbomodules, `getConstants()` will become a regular native module method, and each invocation will hit the native side.

### Callbacks

> Native modules also support a unique kind of argument: a callback. Callbacks are used to pass data from Java to Javascript for asynchronous methods. They can also be used to asynchronously execute JS from the native side.

Reviewer: ```suggestion
Native modules also support a unique kind of argument: a callback. Callbacks are used to pass data from Java to JavaScript for asynchronous methods. They can also be used to asynchronously execute JavaScript from the native side.
```

}
```

### Listening to Lifecycle Events

Listening to the activity's LifeCycle events such as `onResume`, `onPause` etc. is very similar to how we implemented `ActivityEventListener`. The module must implement `LifecycleEventListener`. Then, you need to register a listener in the module's constructor like so:

```jsx
reactContext.addLifecycleEventListener(this);
```

Now you can listen to the activity's LifeCycle events by implementing the following methods:

```jsx
@Override
public void onHostResume() {
   // Activity `onResume`
}
@Override
public void onHostPause() {
   // Activity `onPause`
}
@Override
public void onHostDestroy() {
   // Activity `onDestroy`
}
```

### Threading

To date, on Android, all native module async methods execute on one thread. On iOS, each native module can provide its method queue, or have React Native generate a method queue for it. On iOS, native modules also have an API to get a hold of the method queue on which their async methods are executed.

> Native modules should not have any assumptions about what thread they are being called on, as the current assignment is subject to change in the future. If a blocking call is required, the heavy work should be dispatched to an internally managed worker thread, and any callbacks distributed from there.

Reviewer: It looks like we don't mention the `invalidate` Android analogue. In Java, NativeModules can have a `onCatalystInstanceDestroy()` method that they can implement to do cleanup of any resources they own. There's also an initialize method: https://github.com/facebook/react-native/blob/ae418b83c071d27d1a5bc7dcb46d546686bd21b8/ReactAndroid/src/main/java/com/facebook/react/bridge/NativeModule.java#L40-L51.

Author: @RSNara I ended up removing the notes about iOS here. Do you want to add this info about Android here?
